fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,380 @@
+		Optional<TenantEntryCachePayload<T>> ret = cache->lookupById(tenantId);
+		if (ret.present()) {
+			cache->hits += 1;
+			return ret;
+		}
+
+		TraceEvent(SevInfo, "TenantEntryCacheGetByIdRefresh").detail("TenantId", tenantId);
+
+		// Entry not found. Refresh cacheEntries by scanning underlying KeyRange.
+		// TODO: Cache will implement a "KeyRange" watch, monitoring notification when a new entry gets added or any
+		// existing entry gets updated within the KeyRange of interest. Hence, misses would be very rare
+		wait(refreshImpl(cache, TenantEntryCacheRefreshReason::CACHE_MISS));
+
+		cache->misses += 1;
+		return cache->lookupById(tenantId);
+	}
+
+	ACTOR static Future<Optional<TenantEntryCachePayload<T>>> getByNameImpl(TenantEntryCache<T>* cache,
+	                                                                        TenantName name) {
+		Optional<TenantEntryCachePayload<T>> ret = cache->lookupByName(name);
+		if (ret.present()) {
+			cache->hits += 1;
+			return ret;
+		}
+
+		TraceEvent(SevInfo, "TenantEntryCacheGetByNameRefresh").detail("TenantName", name.contents().toString());
+
+		// Entry not found. Refresh cacheEntries by scanning underlying KeyRange.
+		// TODO: Cache will implement a "KeyRange" watch, monitoring notification when a new entry gets added or any
+		// existing entry gets updated within the KeyRange of interest. Hence, misses would be very rare
+		wait(refreshImpl(cache, TenantEntryCacheRefreshReason::CACHE_MISS));
+
+		cache->misses += 1;
+		return cache->lookupByName(name);
+	}
+
+	Optional<TenantEntryCachePayload<T>> lookupById(int64_t tenantId) {
+		Optional<TenantEntryCachePayload<T>> ret;
+		auto itr = mapByTenantId.find(tenantId);
+		if (itr == mapByTenantId.end()) {
+			return ret;
+		}
+
+		return itr->value;
+	}
+
+	Optional<TenantEntryCachePayload<T>> lookupByName(TenantName name) {
+		Optional<TenantEntryCachePayload<T>> ret;
+		auto itr = mapByTenantName.find(name);
+		if (itr == mapByTenantName.end()) {
+			return ret;
+		}
+
+		return itr->value;
+	}
+
+	Future<Void> refresh(TenantEntryCacheRefreshReason reason) { return refreshImpl(this, reason); }
+
+	static TenantEntryCachePayload<Void> defaultCreatePayload(const TenantName& name, const TenantMapEntry& entry) {
+		TenantEntryCachePayload<Void> payload;
+		payload.name = name;
+		payload.entry = entry;
+
+		return payload;
+	}
+
+	Future<Void> removeEntryInt(Optional<int64_t> tenantId,
+	                            Optional<KeyRef> tenantPrefix,
+	                            Optional<TenantName> tenantName,
+	                            bool refreshCache) {
+		typename Map<int64_t, TenantEntryCachePayload<T>>::iterator itrId;
+		typename Map<TenantName, TenantEntryCachePayload<T>>::iterator itrName;
+
+		if (tenantId.present() || tenantPrefix.present()) {
+			// Ensure either tenantId OR tenantPrefix is valid (but not both)
+			ASSERT(tenantId.present() ^ tenantPrefix.present());
```

Comment:
Nit: `^` is not a logical operator, so while this works it may not be as clear as the following:

```suggestion
			ASSERT(tenantId.present() != tenantPrefix.present());
```

Reply:
Goal is to make sure only one condition is true; for instance: the routine wants either: tenantId or tenantPrefix to be present (and `not` both).

Reply:
That's right, and my proposal does that as well. It's basically the logical operator for xor with two booleans, whereas `^` is an arithmetic operator.

Reply:
I might be missing something, as the suggestion above is replace `^` -> `!=` ?:)

Reply:
For boolean `x` and `y`, `x != y` will be true if either `x` or `y` is true, but not if both are true. `x ^ y` has the same property, but I think this happens indirectly by conversion through integers and requires a deeper understanding of what is happening to know that it is correct. This is because it is not doing a logical operation but rather doing a bitwise operation on the integer representations of true and false.

Reply:
Actually that makes sense, as!= contract is both x & y are not the same value. So both true and/or false are already covered. I was thinking more interms of XOR table covering this usecase, hence, the usage. 

Thanks for pointing out.